### ISSUE TITLE
[ErrorException] Undefined variable: nameMatchesRequiredPackage when running "composer update --lock"

### ISSUE THREAD

**Issue by tonix-tuft** [composer/composer#7933]
My `composer.json`:

```json
{
    "require": {
        "thiagoalessio/tesseract_ocr": "1.1.0",
        "box/spout": "^2.7",
        "sabas/edifact": "^0.3.2",
        "php-edifact/edifact-mapping": "^0.3.2",
        "hackzilla/password-generator": "^1.4",
        "spatie/pdf-to-text": "^1.1"
    },
    "config": {
        "optimize-autoloader": true
    }
}

```

Output of `composer diagnose`:

```
Checking composer.json: WARNING
No license specified, it is recommended to do so. For closed-source software you may use "proprietary" as license.
require.thiagoalessio/tesseract_ocr : exact version constraints (1.1.0) should be avoided if the package follows semantic versioning
Checking platform settings: OK
Checking git settings: OK
Checking http connectivity to packagist: OK
Checking https connectivity to packagist: OK
Checking github.com rate limit: OK
Checking disk free space: OK
Checking pubkeys: 
Tags Public Key Fingerprint: 57815BA2 7E54DC31 7ECC7CC5 573090D0  87719BA6 8F3BB723 4E5D42D0 84A14642
Dev Public Key Fingerprint: 4AC45767 E5EC2265 2F0C1167 CBBB8A2B  0C708369 153E328C AD90147D AFE50952
OK
Checking composer version: OK
Composer version: 1.8.2
PHP version: 7.0.22
PHP binary path: /Applications/MAMP/bin/php/php7.0.22/bin/php

```

When I run this command:

```
composer update --lock
```

I get the following output:

```

                                                  
  [ErrorException]                                
  Undefined variable: nameMatchesRequiredPackage  
                                                  

update [--prefer-source] [--prefer-dist] [--dry-run] [--dev] [--no-dev] [--lock] [--no-custom-installers] [--no-autoloader] [--no-scripts] [--no-progress] [--no-suggest] [--with-dependencies] [--with-all-dependencies] [-v|vv|vvv|--verbose] [-o|--optimize-autoloader] [-a|--classmap-authoritative] [--apcu-autoloader] [--ignore-platform-reqs] [--prefer-stable] [--prefer-lowest] [-i|--interactive] [--root-reqs] [--] [<packages>]...
```

And I expected this to happen:

I expected it to run smoothly and not to return me an error with exit code 1. I recently upgraded macOS from Sierra to High Sierra, before I never experienced this error, maybe that could be the isse. How could I fix it?
I also tried to reinstall composer, but nothing.

PHP version: 7.0.22
OS: macOS High Sierra 10.13.6
Composer: Composer version 1.8.2 2019-01-29 15:00:53

Thank you for the support!

**Comment by tonix-tuft** [comment]
Fiuuu, glad I'm not the only one. I hope so, I checked the code and maybe there could be cases when the variable is actually undefined:

![screen shot 2019-01-29 at 19 05 06](https://user-images.githubusercontent.com/10134421/51929614-edcce300-23f8-11e9-8d85-e531b3ec61a0.png)

Is someone else experiencing the same problem but with a version different from `1.8.2`?

**Comment by ericpoe** [comment]
I experienced this in 1.8.2 and rolled back to 1.8.0. Am not experiencing this in 1.8.0.

**Comment by tonix-tuft** [comment]
Sounds like a bug of 1.8.2 then...

**Comment by paslandau** [comment]
So did I. Luckily you can pass a version to the installer (see https://getcomposer.org/download/ )

My script now looks like this:

```
COMPOSER_VERSION=" --version=1.8.0"
COMPOSER_HASH="48e3236262b34d30969dca3c37281b3b4bbe3221bda826ac6a9a62d6444cdb0dcd0615698a5cbe587c3f0fe57a54d8f5"
php -r "copy('https://getcomposer.org/installer', 'composer-setup.php');" && \
php -r "if (hash_file('sha384', 'composer-setup.php') === '$COMPOSER_HASH') { echo 'Installer verified'; } else { echo 'Installer corrupt'; unlink('composer-setup.php'); } echo PHP_EOL;" && \
php composer-setup.php $COMPOSER_VERSION && \
mv composer.phar /usr/bin/composer && \
php -r "unlink('composer-setup.php');" \
;

# update removed until "Undefined variable: nameMatchesRequiredPackage" -- https://github.com/composer/composer/issues/7933 is fixed
#composer self-update
```

**Comment by tonix-tuft** [comment]
I did the same. Will wait for 1.8.2 to be fixed before upgrading. Thank you guys.

**Comment by hco** [comment]
Just FYI, I created #7932 which could fix this.

**Comment by Seldaek** [comment]
Fixed in 1.8.3
